src/transformers/models/maskformer/modeling_maskformer.py

```diff
@@ -0,0 +1,1638 @@
+            cost_matrix: Tensor = (
+                self.cost_mask * cost_mask + self.cost_class * cost_class + self.cost_dice * cost_dice
+            )
+            # do the assigmented using the hungarian algorithm in scipy
+            assigned_indices: Tuple[np.array] = linear_sum_assignment(cost_matrix.cpu())
+            indices.append(assigned_indices)
+
+        # TODO this is a little weird, they can be stacked in one tensor
+        matched_indices = [
+            (torch.as_tensor(i, dtype=torch.int64), torch.as_tensor(j, dtype=torch.int64)) for i, j in indices
+        ]
+        return matched_indices
+
+    def __repr__(self):
+        head = "Matcher " + self.__class__.__name__
+        body = [
+            "cost_class: {}".format(self.cost_class),
+            "cost_mask: {}".format(self.cost_mask),
+            "cost_dice: {}".format(self.cost_dice),
+        ]
+        _repr_indent = 4
+        lines = [head] + [" " * _repr_indent + line for line in body]
+        return "\n".join(lines)
+
+
+# copied from original implementation
+class MaskFormerLoss(nn.Module):
+    def __init__(
+        self,
+        num_classes: int,
+        matcher: MaskFormerHungarianMatcher,
+        weight_dict: Dict[str, float],
+        eos_coef: float,
+        losses: List[str],
+    ):
+        """The MaskFormer Loss. The loss is computed very similar to DETR. The process happens in two steps:
+        1) we compute hungarian assignment between ground truth masks and the outputs of the model
+        2) we supervise each pair of matched ground-truth / prediction (supervise class and mask)
+
+        Args:
+            num_classes (int): The number of classes
+            matcher (MaskFormerHungarianMatcher): A torch module that computes the assigments between the predictions and targets
+            weight_dict (Dict[str, float]): A dictionary of weights to be applied to the different losses
+            eos_coef (float): TODO no idea
+            losses (List[str]): A list of losses to be used TODO probably remove it
+        """
+
+        super().__init__()
+        requires_backends(self, ["scipy"])
+        self.num_classes = num_classes
+        self.matcher = matcher
+        self.weight_dict = weight_dict
+        self.eos_coef = eos_coef
+        self.losses = losses
+        empty_weight: Tensor = torch.ones(self.num_classes + 1)
+        empty_weight[-1] = self.eos_coef
+        self.register_buffer("empty_weight", empty_weight)
+
+    def loss_labels(
+        self, outputs: Dict[str, Tensor], targets: Dict[str, Tensor], indices: Tuple[np.array], num_masks: float
+    ) -> Dict[str, Tensor]:
+        """Classification loss (NLL)
+        # TODO this doc was copied by the authors
+        targets dicts must contain the key "labels" containing a tensor of dim [nb_target_masks]
+        """
+        assert "pred_logits" in outputs
+        pred_logits: Tensor = outputs["pred_logits"]
+        b, q, _ = pred_logits.shape
```

Comment:
Better variable names please.

Reply:
`b` and `q` were picked for consistency since they represent the same shape that is used later in `.rearrange`, do you think I should change the name there?

Reply:
One-letter variable names are not readable, and as said before we should drop the requirement on einops anyway, so `rearrange` will go away.